drivers/FlashIAP.cpp

```diff
@@ -0,0 +1,129 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "FlashIAP.h"
+#include "mbed_assert.h"
+
+#ifdef DEVICE_FLASH
+
+namespace mbed {
+
+FlashIAP::FlashIAP()
+{
+
+}
+
+FlashIAP::~FlashIAP()
+{
+
+}
+
+int FlashIAP::init()
+{
+    if (flash_init(&_flash)) {
+        return -1; 
+    } else {
+        return 0;
+    }
+}
+
+int FlashIAP::deinit()
+{
+    if (flash_free(&_flash)) {
+        return -1; 
+    } else {
+        return 0;
+    }
+}
+
+
+int FlashIAP::read(void *buffer, uint32_t addr, uint32_t size)
+{
+    uint32_t count = size;
+    uint8_t *flash_pointer = (uint8_t *)addr;
+    uint8_t *read_buffer = (uint8_t *)buffer;
+    while (count) {
+        *read_buffer = *flash_pointer;
+        flash_pointer++;
+        read_buffer++;
+        count--;
+    }
+    return 0;
+}
+
+int FlashIAP::program(const void *buffer, uint32_t addr, uint32_t size)
+{
+    if (flash_program_page(&_flash, addr, (const uint8_t *)buffer, size)) {
+        return -1;
+    } else {
+        return 0;
+    }
+}
+
+int FlashIAP::write(const void *buffer, uint32_t addr, uint32_t size)
+{
+    if (erase(addr, size) != 0) {
+        return -1;
+    }
+
+    if (flash_program_page(&_flash, addr, (const uint8_t *)buffer, size)) {
+        return -1;
+    } else {
+        return 0;
+    }
+}
+
+int FlashIAP::erase(uint32_t addr, uint32_t size)
+{
+    MBED_ASSERT((flash_get_sector_size(&_flash, addr) % size) == 0);
```

Comment:
Should probably also check alignment? `(addr / sector_size) * sector_size == addr`